### ISSUE TITLE
Handle re-installation of 3rd party app

### ISSUE THREAD

**Issue by tobiasKaminsky** [nextcloud/Android-SingleSignOn#12]
- install 3rd party app
- allow access
- re-install 3rd party app
- token is already stored in files app (as it is based on package name), but 3rd party app has no longer the token
--> we need some way to re-authenticate.

**Comment by tobiasKaminsky** [comment]
- [ ] remove 3rd party app after everything worked & re-install it --> should ask again for authentication
- [ ] remove files app after everything worked & re-install it -> 3rd party also should again ask for authentication

**Comment by tobiasKaminsky** [comment]
First one works out of the box as we overwrite token on "grant access".

**Comment by tobiasKaminsky** [comment]
Second one returns "invalid token", which is in https://github.com/nextcloud/android/blob/c52f9c842f44b5c9e42b1125468fa03dd554d162/src/main/java/com/nextcloud/android/sso/Constants.java#L17

@stefan-niedermann @desperateCoder 3rd party apps should check for this and start their own re-authenticate process again.

Would that work for you?

**Comment by desperateCoder** [comment]
We could do it like this, but should we? This is a case, future users of the SSO have to be aware of. Is it impossible for the SSO lib to handle this?

For me this shouldn't be a problem, but it's inconvenient tbh.

**Comment by tobiasKaminsky** [comment]
hm, I have no idea how to handle this via SSO:
- install files app & authenticate account
- install 3rd party app
- grant account to 3rd party
	- token is stored in files app (this is needed to verify app calls)
	- sso account is stored via shared preferences in 3rd party app
- remove files app --> no tokens are available
- re-install files app
- login into files app with account from above
- 3rd party app tries to access said account
--> files app will refuse as there is no token in files app

tl;dr: if files app is deleted, we do not know anything about already authenticated accounts/apps.

**Comment by desperateCoder** [comment]
So how can you tell, when to throw the exception mentioned above? Isn't that exactly the case when we have to reauthentificate?

**Comment by tobiasKaminsky** [comment]
The exception is thrown by SSO/files.
I'll change this in above scenario.

**Comment by desperateCoder** [comment]
So now I'm confused. File app says CE_1. SSO still knows selected account. Can't SSO catch CE_1 and try to reauthentificate/re-grant?

Sorry if I am talking nonsense, I don't have had a deep-dive into the whole process.

**Comment by AndyScherzinger** [comment]
@desperateCoder I just build the latest master for the SNAPSHOT mechanism of jitpack to pick it up in case you build/compile the deck app: https://jitpack.io/com/github/nextcloud/Android-SingleSignOn/master-8b590d8028-1/build.log

**Comment by desperateCoder** [comment]
@AndyScherzinger ok? afaik i currently use -SNAPSHOT as version for SSO, but i wonder if all problems are already solved? See above... :thinking:

**Comment by AndyScherzinger** [comment]
@desperateCoder not sure what has been fixed already...

**Comment by David-Development** [comment]
Okay, just pushed some changes. If you want to handle reauthentication, you can use this snippet

```java
// caught TokenMismatchException
try {
    SingleAccountHelper.reauthenticateCurrentAccount(this);
} catch (NextcloudFilesAppAccountNotFoundException | NoCurrentAccountSelectedException | NextcloudFilesAppNotSupportedException e) {
    UiExceptionManager.showDialogForException(this, e);
} catch (NextcloudFilesAppAccountPermissionNotGrantedException e) {
    // Unable to reauthenticate account just like that..
    // TODO Show login screen here
}
```

Example: https://github.com/nextcloud/news-android/blob/master/News-Android-App/src/main/java/de/luhmer/owncloudnewsreader/NewsReaderListActivity.java#L435

**Comment by tobiasKaminsky** [comment]
> So now I'm confused. File app says CE_1. 

File app says CE_1 because Files app is storing the token and after re-installation the token is gone.

> SSO still knows selected account. 

There is no "SSO". Just think of SSO as your app: SSO is storing infos about known account / token within your shared config of your app.
That is the reason why "SSO"/your app still knows the account/token, because it is not deleted.

> Can't SSO catch CE_1 and try to reauthentificate/re-grant?

See above. SSO is part of your app and like the initial account setup you have to react on it. SSO never knows how you implemented your first login.
(with David's change it is hopefully now a bit more cleaner)

**Comment by desperateCoder** [comment]
Ok, I see. It's ok as it is and as @david-development proposed.

I see the difficulties but I'm sure, we could solve this with a strategy pattern if we wanted. Just by passing the behavior by implementing an interface. 

Don't get me wrong, this is fine tuning, nice to have, I'm just pitching.

**Comment by tobiasKaminsky** [comment]
so, this can be closed, or? :-)
